pandas/_libs/src/vendored/ujson/python/objToJSON.c

```diff
@@ -1303,35 +1314,39 @@ char **NpyArr_encodeLabels(PyArrayObject *labels, PyObjectEncoder *enc,
                              "Cannot cast numpy dtype %d to long",
                              enc->npyType);
             }
-            castfunc(dataptr, &nanosecVal, 1, NULL, NULL);
+            castfunc(dataptr, &i8date, 1, NULL, NULL);
+            // copied from
+            // https://github.com/numpy/numpy/blob/c8fe278a754a271af57eaf6c7ffb2382e5a954f9/numpy/core/src/multiarray/datetime.c#L692-L701
+            dateUnit = ((PyArray_DatetimeDTypeMetaData *)dtype->c_metadata)->meta.base;
```

Comment:
This part is probably the most sketchy part of this PR.

We currently do this the exact same way numpy does it, but not sure if this is part of the public numpy C API.

FWIW, ``PyArray_DatetimeDTypeMetaData`` appears in ``ndarraytypes.h``.

Reply:
It would be better if we stuck to a method exposed in the datetime header. Does `get_datetime_metadata_from_dtype` not help here?

Reply:
It would be nice to use that, not sure if that's public either (it's defined in ``multiarray/_datetime.h``).
Like ``PyArray_DatetimeDTypeMetaData``, it also seems to be undocumented.

Maybe one of the numpy maintainers(@seberg?) can clarify more?